sled-agent/src/common/instance.rs

```diff
@@ -288,6 +224,55 @@ impl InstanceStates {
         &mut self,
         observed: &ObservedPropolisState,
     ) -> Option<Action> {
+        fn transition_migration(
+            current: &mut Option<MigrationRuntimeState>,
+            ObservedMigrationState { id, state }: ObservedMigrationState,
+            now: DateTime<Utc>,
+        ) {
+            if let Some(ref mut m) = current {
+                // Don't generate spurious state updates if the migration is already in
+                // the state we're transitioning to.
+                if m.migration_id == id && m.state == state {
+                    return;
+                }
+                m.state = state;
+                if m.migration_id == id {
+                    m.gen = m.gen.next();
+                } else {
+                    m.migration_id = id;
+                    m.gen = Generation::new().next();
+                }
+                m.time_updated = now;
+            } else {
+                *current = Some(MigrationRuntimeState {
+                    migration_id: id,
+                    // We are creating a new migration record, but the state
+                    // will not be `Pending`, because we've actually gotten a
+                    // migration observation from Propolis. Therefore, we have
+                    // to advance the initial generation once to be ahead of
+                    // what the generation in the database is when Nexus creates
+                    // the initial migration record at generation 1.
+                    gen: Generation::new().next(),
+                    state,
+                    time_updated: now,
+                });
+            }
+        }
+
+        fn destroy_migration(
+            migration: &mut MigrationRuntimeState,
+            now: DateTime<Utc>,
+        ) {
+            if matches!(
+                migration.state,
+                MigrationState::InProgress | MigrationState::Pending
+            ) {
```

Comment:
Do you think `!migration.is_terminal()` would be a safer check here? Just trying to future-proof against a new `MigrationState` variant

Reply:
Yeah, I think this was written before `migration.is_terminal()` existed --- good call!